### ISSUE TITLE
Jak Collection/inFamous games Broken Refraction

### ISSUE THREAD

**Issue by JimScript** [RPCS3/rpcs3#11149]
Throughout the Jak collection, the particles that refract the scene are broken, even though other refracting effects like water and the scattershot work fine. Some parts of the particles mesh with the scene which leads me to believe this is a uv problem.

Jak and Daxter:
![BCUS98281_screenshot_2021_11_12_11_48_23](https://user-images.githubusercontent.com/66685584/141527668-c560ede5-1ade-444a-95f1-93e6a6357e08.png)
[BCUS98281_20211112114854_capture.zip](https://github.com/RPCS3/rpcs3/files/7530134/BCUS98281_20211112114854_capture.zip)

Jak II:
![BCUS98281_screenshot_2021_11_12_11_51_28](https://user-images.githubusercontent.com/66685584/141527752-b93404d4-810f-442d-8986-0209b905e596.png)
[BCUS98281_20211112115137_capture.zip](https://github.com/RPCS3/rpcs3/files/7530137/BCUS98281_20211112115137_capture.zip)

System:
- SYS: RPCS3 v0.0.19-13006-8b3ed7b5 Alpha | HEAD
- SYS: AMD Ryzen 5 3600 6-Core Processor | 12 Threads | 15.95 GiB RAM | TSC: 3.600GHz | AVX+ | FMA3
- SYS: Operating system: Windows, Major: 10, Minor: 0, Build: 19043, Service Pack: none, Compatibility mode: 0
- RSX: Found vulkan-compatible GPU: 'NVIDIA GeForce RTX 2070 SUPER' running on driver 471.68.0.0
- SYS: Firmware version: 4.85

**Comment by Zorklis** [comment]
Yep, it's pretty noticable in the Jak 3 intro cutscene 
https://youtu.be/0pRW9OlcIGI

**Comment by JimScript** [comment]
I tried to get a photo of a real ps3, but when the effect is working correctly it is hard to see.
![0405221653_HDR 1](https://user-images.githubusercontent.com/66685584/161870513-99b09275-e8c8-47fb-88ec-2dad37ad1b61.jpg)
![0405221653a_HDR 1](https://user-images.githubusercontent.com/66685584/161870547-572c9882-5cc9-4e8c-80bd-f25b56a072a3.jpg)

**Comment by kd-11** [comment]
Maybe you can circle or highlight the difference? I have no idea what this is about otherwise (never played jak)

**Comment by MSuih** [comment]
The Jak II scene is supposed to be an air vent blasting hot air out. In his last image you can see it slightly distort the image. In the first post the same image looks like some ice cubes were stacked on top of the vent or something.

**Comment by JimScript** [comment]
And because of that it is hard to put a red circle around it, it is basically the warped image behind jak above the vent. What it is suppose to look like is a smudge like effect, like the image is bulging out.

**Comment by kd-11** [comment]
Oh ok, so its just the infamous distortion shader glitch.
For deformation shaders the UVs are from 2 sources, fixed UV in screen space and a deformation vector usually gleaned from dFdx and dFdy. The functionality is implemented but doesn't work right. I haven't had time to look into the distortion issues, they're fairly low priority all things considered, but if more games are affected I can move it up the queue. Currently I've only heard of it in infamous and now Jak games.

**Comment by RinMaru** [comment]
> (never played jak)
shame on you >.< 

but yea i get this error too. ill be upgrading from RX 480 to RX 6700XT soon so i can test on that it that would help with figuring it out

**Comment by Asinin3** [comment]
FYI, One other game I saw someone have this issue in was Persona 5, https://github.com/RPCS3/rpcs3/issues/6993 but I couldn't reproduce it and they didn't update the issue so it was closed.

**Comment by JimScript** [comment]
Is the Persona 5 issue the same, because in both inFamous and Jak the image in the shader is offset to the left and up a bit. Here is a picture of it in action in inFamous:
![BCUS98119_screenshot_2022_04_08_18_39_17](https://user-images.githubusercontent.com/66685584/162551648-1af49df7-95fa-48d9-9c1c-91cfc7757dab.png)

You can also see it in the Jak II image and a little bit in the Jak I image above, but it is hard to see in the Persona 5 examples in that thread.

**Comment by Darkhost1999** [comment]
![BF3](https://user-images.githubusercontent.com/60384196/164916387-05851447-cc03-4fca-8741-651c963fa698.jpg)
Fires, smoke and rain all have these squares/distortion around them when you can get a clear image from this unstable game.
So I think we can add BF3 to the list?

**Comment by JimScript** [comment]
Do you have any examples of fire distortion being broken? It is a little hard to see and the blocking kinda reminds me of anti-aliasing breaking.

**Comment by Darkhost1999** [comment]
> Do you have any examples of fire distortion being broken? It is a little hard to see and the blocking kinda reminds me of anti-aliasing breaking.  

https://youtu.be/WhkFCPfA5jg
![Screenshot_20220423-161335.png](https://user-images.githubusercontent.com/60384196/164946153-184f0875-72e8-4e0e-9b5c-7fea6b52acdb.png)

There's quite a lot of graphical issues it's possible the cause is different. But when looking at burning tanks on fire and smoking with minimal graphics issues it looks like infamous's distortion.

**Comment by JimScript** [comment]
Wow that is _a lot_ more broken than your first post implied, but looking at it closely it looks like MLAA or something is broken. If you look around 4:45 you see the squares form around the flagpoles and the edge of the sculpture. If this error is in effect too, I can hardly tell.

**Comment by Darkhost1999** [comment]
Hopefully me stopping to take in the scenery helps in Thunder run
https://user-images.githubusercontent.com/60384196/164947136-fc1f0298-7db2-4e1c-88df-b95aa1d56f23.mp4

**Comment by JimScript** [comment]
Yeah it kinda looks like MLAA is drawing over the games particle systems. Jak and inFamous are both ingame shader problems, this looks like a post-processing problem.

**Comment by Darkhost1999** [comment]
> Yeah it kinda looks like MLAA or MSAA is drawing over the games particle systems. Jak and inFamous are both ingame shader problems, this looks like a post-processing problem.

![image](https://user-images.githubusercontent.com/60384196/164948035-67d25502-b197-4414-a7bd-95893e63c12e.png)
It'd make more sense if the issue went away disabling this and RPCS3 AA.
This being disabled in all videos and screenshots above. RPCS3 being on.

**Comment by JimScript** [comment]
Well now that is strange, maybe it is the same issue, maybe it's just that Battlefield 3 uses the function in a different way than Jak or inFamous. But it is hard to see the similarities, as both Jak and inFamous have it as a glassy image that is offset to left. What makes you think that this is the same issue?

**Comment by Darkhost1999** [comment]
> What makes you think that this is the same issue?

Lack of better knowledge ¯\_(ツ)_/¯
And how it appears to me.

**Comment by JimScript** [comment]
Well I guess we won't know for sure if this or the Persona 5 issue is related until it is fixed.

**Comment by Codex-** [comment]
It sounds like you already have a strong suspicion as to what the issue is, kd-11, I was having a read through the jak and daxter decompilation project and thought it might help provide some more concrete insight to the implementation details for these. Albeit for the ps2 version but who knows: https://github.com/open-goal/jak-project/tree/master/goal_src/engine

Here's to hoping it helps a little 😅

**Comment by JimScript** [comment]
Well I imagine that Mass*Media(the company that ported the collection) might have changed a few things behind the scenes for it to work on the PS3. But then again, the Jak project could have too. The PS2 and PS3 are two _very_ different consoles, so the work might not be transferable.

**Comment by cRAzyyKirill0101** [comment]
> Проблема с Persona 5 такая же, потому что и в inFamous, и в Jak изображение в шейдере смещено влево и немного вверх. Вот изображение этого в действии в InFamous: ![BCUS98119_screenshot_2022_04_08_18_39_17](https://user-images.githubusercontent.com/66685584/162551648-1af49df7-95fa-48d9-9c1c-91cfc7757dab.png)
> 
> Вы также можете увидеть это на изображении Jak II и немного на изображении Jak I выше, но это трудно увидеть в примерах Persona 5 в этой ветке.

hi, I need help how did you launch this game on your system, what characteristics do you have, and can you tell what configuration you have in the emulator, I just run it and I have 5-10 frames per second, I can't understand why the gpu and cpu don't go to the ceiling when "playing" here I think because of what can it be, and yes, sorry if that offtop just really need help because I really love this game ((((

**Comment by JimScript** [comment]
Thanks for bumping the issue, maybe I should change the name to include inFamous 1 so these issues stop popping up.

**Comment by JimScript** [comment]
For some reason, inFamous 1 fades into the broken refraction and at a certain distance some of it is fine, with it getting progressively worse when it gets closer to the camera:


https://user-images.githubusercontent.com/66685584/214114614-274b919f-a900-4680-94c4-d8b3f4b8ac83.mp4

**Comment by Llamamoto** [comment]
I've got $10 in BTC for whoever can fix this;)

**Comment by JimScript** [comment]
![infamousRenderExample](https://user-images.githubusercontent.com/66685584/215350181-cb0d06ce-b82f-40a8-805e-8f53ac21ee85.png)
Okay, so I tried dissecting the scene and I found this normal that dictates the refraction:
![infamousNormalExample](https://user-images.githubusercontent.com/66685584/215349150-02b5e74d-69bf-4126-87e8-d6c828289edf.png)
I then dissected it further to see what the problem was, and found that the working parts are all increments from the 128 base value while the broken parts are all solidly 255.
Here is an example I made for the Red Channel, exaggerated for easy viewing:
![infamousRedExample](https://user-images.githubusercontent.com/66685584/215349312-67fdb4aa-43fb-49fb-b052-0b1c8886cbbe.png)
The red parts go from 129-131 red while the blue parts are all 255 red, interesting that they are on opposite sides. Which makes me think there might be an integer underflow problem when it tries to become 127 or lower, so it loops back to 255.
Here is the example for Green Channel too, also exaggerated:
![infamousGreenExample](https://user-images.githubusercontent.com/66685584/215349541-ba30b19e-a394-4496-9123-6fb577e6634a.png)
Don't know how this affects Jak though.
Here is the RSX capture of the scene:
[BCUS98119_20230128235239_capture.zip](https://github.com/RPCS3/rpcs3/files/10530617/BCUS98119_20230128235239_capture.zip)

**Comment by Hichem47** [comment]
someone can fix this issue please, thanks

**Comment by JimScript** [comment]
Well, I decided to look back into how the Jak issue relates to the inFamous issue:
![JakRenderExample](https://github.com/RPCS3/rpcs3/assets/66685584/b593b882-a960-4bbf-82d1-537f5262f64f)
With the normal that the refraction uses of course:
![JakNormalExample](https://github.com/RPCS3/rpcs3/assets/66685584/c0ae1eba-c071-4b5d-b21f-4f20bbc8cc2a)
And it's similar in that below its midway point it whites out, but this midway point is 127 instead. It also looks like it's more based off of polygons instead of textures, and when a any pixel in a triangle goes below 127, then that whole triangle for that channel whites out. That's probably why some of the polygons flicker ingame, they're going in and out of the broken range. Here are the channels re-rendered in between 127-140 for better viewing:
Red Example:
![JakRedExample](https://github.com/RPCS3/rpcs3/assets/66685584/3f51f7d2-cd48-4a85-ad1b-04a834bf468a)
Green Example:
![JakGreenExample](https://github.com/RPCS3/rpcs3/assets/66685584/4a84fde5-14a4-434b-8e23-69240a1be027)

**Comment by matthewmgraf** [comment]
I would love to learn how to fix this, whether it means I have to learn some game design or whatever, but do we have any leads on how to fix this issue?

**Comment by KieronJ** [comment]
For reference, this shader works by rendering the pentagons with additive blending. The vertex "color" seen above comes from vert_uv - vert_xy. A second pass then uses the RG channels of the pentagon buffer as a screen space displacement.

**Comment by Ronin1mil** [comment]
Recently found this thread and would like to add that this also happens in infamous 2. It's extremely noticeable at the start of the game. It could be my setup, but I doubt it. I'm going to try infamous festival of blood next, just to see if it works properly or not. First image is what it's supposed to look like, while the second is gameplay I stole form youtube. 
![Screenshot 2023-12-31 113150](https://github.com/RPCS3/rpcs3/assets/95828084/c6b03729-9d8e-48ec-8917-94bb628eae72)
![Screenshot 2023-12-31 113414](https://github.com/RPCS3/rpcs3/assets/95828084/e2c3d33e-fe7a-49ff-8a0a-7c33a5687086)

**Comment by Pazifix** [comment]
Any new leads or updates on fixing this issue?

**Comment by AlexS2412** [comment]
still nothing i guess years later? nobody cared enough to fix this in jak or infamous... tbh i feel like just modding the games to remove the effect as a work around

**Comment by Pazifix** [comment]
> still nothing i guess years later? nobody cared enough to fix this in jak or infamous... tbh i feel like just modding the games to remove the effect as a work around

Yeah really a shame cause jak and infamous are true gems and deserve to be played in its full glory, I would help but I am a noob on this subject, I just really like infamous and jak.

**Comment by AniLeo** [comment]
I'm locking this due to all the off topic spam, if you have new information to add ping me on Discord to have it reopened, otherwise kd will reopen it when he works on it

**Comment by Hasster1** [comment]
Some more info from the InFamous' side of things:
While trying my luck at finding some interesting stuff with Cheat Engine, i've stumbled upon one opcode, that re-activated the bug with the glitched borders of the screen, that was fixed in https://github.com/RPCS3/rpcs3/pull/4973
I also noticed, that with that bug present on the screen, the broken reflections did not show up (The fire usually produced the reflection particles, but they were hidden)
I've traced that opcode in ghidra, and found the value, that controlled the amount of shifting for both particles and the glitched border. Default amount is set to 0.1.

I hope this info helps a bit with fixing this bug.

**Comment by AlexS2412** [comment]
> Some more info from the InFamous' side of things: While trying my luck at finding some interesting stuff with Cheat Engine, i've stumbled upon one opcode, that re-activated the bug with the glitched borders of the screen, that was fixed in #4973 I also noticed, that with that bug present on the screen, the broken reflections did not show up (The fire usually produced the reflection particles, but they were hidden) I've traced that opcode in ghidra, and found the value, that controlled the amount of shifting for both particles and the glitched border. Default amount is set to 0.1.
> 
> I hope this info helps a bit with fixing this bug.

can you produce a test build with this workaround? would be much appreciated.

**Comment by Hasster1** [comment]
> can you produce a test build with this workaround? would be much appreciated.

There's already a patch for InFamous that sets the shifting to 0, disabling the effect.

**Comment by DarkPhobia0** [comment]
Just to understand, this doesn’t fix the effect, just disables it, right?

Thanks!

**Comment by Hasster1** [comment]
> Just to understand, this doesn’t fix the effect, just disables it, right? 

Yes, that's right.

**Comment by AlexS2412** [comment]
> > can you produce a test build with this workaround? would be much appreciated.
> 
> There's already a patch for InFamous that sets the shifting to 0, disabling the effect.

can you please share that?

**Comment by Hasster1** [comment]
> can you please share that?

It's... already on the wiki, and rpcs3's patch updater automatically downloads it.
If you've got any more questions not related to the bug itself - please ask them somewhere else.

**Comment by AlexS2412** [comment]
> > can you please share that?
> 
> It's... already on the wiki, and rpcs3's patch updater automatically downloads it. If you've got any more questions not related to the bug itself - please ask them somewhere else.

I've never used the patch updater before... im just lost sorry.

**Comment by Hasster1** [comment]
> I've never used the patch updater before... im just lost sorry.

https://discord.gg/RPCS3

**Comment by AlexS2412** [comment]
> > I've never used the patch updater before... im just lost sorry.
> 
> https://discord.gg/RPCS3

wierdly with the patch applied it doesnt do anything. its wierd.

**Comment by AniLeo** [comment]
Locking the issue again since the spam is back, which is a shame.

**Comment by digant73** [comment]
@JimScript is that still an issue on both games?

**Comment by Virtua64** [comment]
I just tested Jak and yes it is still an issue.

For reference, this is what it looks like now:

RPCS3:

https://github.com/user-attachments/assets/207c2258-0b69-4ba7-9884-3b484788c5cd



PS2 (emu):

https://github.com/user-attachments/assets/f267653d-a38f-414c-a622-77db7e88ab71



It should basically look like what is shown in the ps2 version.

**Comment by JimScript** [comment]
Yep and inFamous still has the same issues too...

**Comment by Camology** [comment]
Smoke effects for infamous 1 look correct now after updating to the latest nightly build, will need to verify with more areas and more games but this may be a solved issue now.

**Comment by JimScript** [comment]
> Smoke effects for infamous 1 look correct now after updating to the latest nightly build, will need to verify with more areas and more games but this may be a solved issue now.

What latest nightly build are you talking about? My RPCS3 still has it broken and I'm on the normal updates, can you give a link to what version you downloaded?

**Comment by Camology** [comment]
https://rpcs3.net/compatibility?b

https://github.com/RPCS3/rpcs3/pull/17764

My updates point to this being the latest nightly build and smoke effects look correct in infamous 1 for me as of now. Update to the latest, 0.0.38-18406 and see if that’s true for you too.

**Comment by JimScript** [comment]
Yeah I already tried it on 0.0.38-18406 and it's still not fixed on my end, are you sure you looked at the smoke close enough? I check by looking closely at the chimney smoke left of Zeke's rooftop. Also are there any special settings you use that might change anything?

**Comment by Camology** [comment]
I just have the default settings from the wiki, it definitely doesn’t look the same for me that it did when I first started the game a few weeks ago. In the middle of an area so cant fast travel back to zekes yet but once I can I’ll take a screenshot and post here, or a video if I can figure out how to.

**Comment by Camology** [comment]
think ill be stuck here for awhile if I remember the game well, so just took a screenshot of some of the smoke around me at the moment. 

https://imgur.com/a/6mxrhyo

Prior this was completely translucent and looked very weird, now completely normal.

**Comment by JimScript** [comment]
I look at the smoke a lot closer as in in the foreground between Cole and the camera to see if it's still there, as it might not appear from that far away. And just to be sure, you don't have the "Disable Broken Refraction" patch on, do you?

**Comment by Camology** [comment]
Ah yep you are right, forgot I had enabled the patch last time. Turned it off and its back to broken again :(  Disregard!

**Comment by digant73** [comment]
probably still an issue but, please re-test on latest build. Hopefully some textures and reflections issues have been fixed by recent build 0.0.39-18790

**Comment by Camology** [comment]
Building a new pc this afternoon, will test again and comment with the results once I have a chance

**Comment by KieronJ** [comment]
> probably still an issue but, please re-test on latest build. Hopefully some textures and reflections issues have been fixed by recent build 0.0.39-18790

Still present as of ed3d2a0.
<img width="1280" height="720" alt="Image" src="https://github.com/user-attachments/assets/808ce3c8-dd8f-4625-a637-239aa3a595b4" />

**Comment by Zin1023** [comment]
Still any fix for the infamous problem?

**Comment by TheXbox** [comment]
When is this getting fixed??? been AGES! it seems the devs forgot about this issue.

**Comment by Jonathan44062** [comment]
Then this thread gets blocked because people like you are impatient. This is not the place to air your grievances. RPCS3 takes time to improve little by little. If you don't have anything interesting to contribute (make a contribution and fix it yourselves), don't say anything and behave yourselves.

**Comment by kd-11** [comment]
We don't even have an updated status after the recent few weeks of fixes that solve issues with reflections and refractions.

**Comment by TheXbox** [comment]
It’s still an issue i can tell you that much

**Comment by Wastelander121** [comment]
> We don't even have an updated status after the recent few weeks of fixes that solve issues with reflections and refractions.

Can confirm it still happens on the latest.

<img width="2562" height="1435" alt="Image" src="https://github.com/user-attachments/assets/2ffb5051-d186-4d5b-a9f5-049392251d62" />

**Comment by treeleaves12** [comment]
I have BCES01325 and tested the first level on Jak1 and the graphical issue is still there, it is very noticeable in action (you can't miss it) although a fairly benign issue. Used the default settings with build 19107. Win11, intel i5-13600k, geforce 3070.

**Comment by ShadowH69** [comment]
People like you are annoying your presence is unwelcome, if you're not gonna add anything constructive and just bitch and complain than you can leave , you pretend like emulation is an easy task , Developers do this in their free time , they still have life's outside of this too , if you want something fixed than either wait until it is or do it yourself.

**Comment by AniLeo** [comment]
The trash has been taken out. Whenever you see spam from disgruntled wastes of oxygen don't engage, just wait for someone from staff to block them.

**Comment by DarkPhobia0** [comment]
Avoiding the noise, team has been addressing lots of visual issues lately,
so hopefully they’ll address this soon!

**Comment by alexconsole04** [comment]
I'm entirely unsure if anyone has mentioned it, but it's not just infamous 1 that has this issue, all the infamous games suffer the same problem.

**Comment by JimScript** [comment]
> I'm entirely unsure if anyone has mentioned it, but it's not just infamous 1 that has this issue, all the infamous games suffer the same problem.

I just checked and yeah inFamous 2 has this problem too, and I'm assuming that Festival of Blood has it as well so maybe I'll change the title to reflect that.
